File: src/api/anomaly.ts
/**
 * Anomaly Detection API Client
 * Real-time telemetry anomaly detection
 */

export interface AnomalyAlert {
  type: 'critical' | 'rate_of_change' | 'ml_detected';
  sensor: string;
  value?: number;
  threshold?: number;
  rate?: number;
  score?: number;
  contributing_features?: string[];
  message: string;
  severity: 'low' | 'medium' | 'high';
}

export interface AnomalyDetectionResult {
  is_anomaly: boolean;
  anomaly_score: number;
  alerts: AnomalyAlert[];
  timestamp: string;
  vehicle_id: string;
  vehicle_number?: number;
  lap?: number;
}

export interface AnomalyStats {
  total_points: number;
  anomaly_count: number;
  anomaly_rate: number;
  critical_alerts: number;
  rate_of_change_alerts: number;
  ml_detected_anomalies: number;
  avg_anomaly_score: number;
  top_anomalous_sensors: Array<{ sensor: string; count: number }>;
}

const API_BASE = import.meta.env.VITE_BACKEND_URL || 'http://localhost:8000';

/**
 * Detect anomalies in a single telemetry point
 */
export async function detectAnomaly(
  vehicleId: string,
  telemetryPoint: Record<string, any>
): Promise<AnomalyDetectionResult> {
  const response = await fetch(`${API_BASE}/api/anomaly/detect`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      vehicle_id: vehicleId,
      telemetry_point: telemetryPoint,
    }),
  });

  if (!response.ok) {
    throw new Error(`Anomaly detection failed: ${response.statusText}`);
  }

  return response.json();
}

/**
 * Get anomaly statistics for a race or vehicle
 */
export async function getAnomalyStats(
  track: string,
  race: number,
  vehicle?: number,
  lapStart?: number,
  lapEnd?: number
): Promise<{ success: boolean; data: AnomalyStats; timestamp: string }> {
  const params = new URLSearchParams({
    track,
    race: race.toString(),
  });

  if (vehicle !== undefined) {
    params.append('vehicle', vehicle.toString());
  }
  if (lapStart !== undefined) {
    params.append('lap_start', lapStart.toString());
  }
  if (lapEnd !== undefined) {
    params.append('lap_end', lapEnd.toString());
  }

  const response = await fetch(`${API_BASE}/api/anomaly/stats?${params}`);

  if (!response.ok) {
    throw new Error(`Failed to get anomaly stats: ${response.statusText}`);
  }

  return response.json();
}

/**
 * Batch anomaly detection
 */
export async function detectAnomaliesBatch(
  vehicleId: string,
  track: string,
  race: number,
  lapStart?: number,
  lapEnd?: number,
  retrain: boolean = true
): Promise<{
  success: boolean;
  stats: {
    total_points: number;
    anomaly_count: number;
    anomaly_rate: number;
    avg_anomaly_score: number;
  };
  results: any[];
  timestamp: string;
}> {
  const response = await fetch(`${API_BASE}/api/anomaly/detect/batch`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      vehicle_id: vehicleId,
      track,
      race,
      lap_start: lapStart,
      lap_end: lapEnd,
      retrain,
    }),
  });

  if (!response.ok) {
    throw new Error(`Batch anomaly detection failed: ${response.statusText}`);
  }

  return response.json();
}

/**
 * Check anomaly detection service health
 */
export async function checkAnomalyHealth(): Promise<{
  status: string;
  pyod_available: boolean;
  active_connections: number;
  timestamp: string;
}> {
  const response = await fetch(`${API_BASE}/api/anomaly/health`);

  if (!response.ok) {
    throw new Error(`Health check failed: ${response.statusText}`);
  }

  return response.json();
}

/**
 * WebSocket client for real-time anomaly alerts
 */
export class AnomalyWebSocket {
  private ws: WebSocket | null = null;
  private vehicleId: string;
  private onAnomaly?: (result: AnomalyDetectionResult) => void;
  private onError?: (error: Event) => void;
  private onConnect?: () => void;
  private onDisconnect?: () => void;

  constructor(vehicleId: string) {
    this.vehicleId = vehicleId;
  }

  connect(
    onAnomaly: (result: AnomalyDetectionResult) => void,
    onError?: (error: Event) => void,
    onConnect?: () => void,
    onDisconnect?: () => void
  ) {
    this.onAnomaly = onAnomaly;
    this.onError = onError;
    this.onConnect = onConnect;
    this.onDisconnect = onDisconnect;

    const wsUrl = import.meta.env.VITE_BACKEND_WS_URL || 'ws://localhost:8000';
    const url = `${wsUrl}/api/anomaly/ws/${this.vehicleId}`;

    this.ws = new WebSocket(url);

    this.ws.onopen = () => {
      console.log(`Anomaly WebSocket connected for vehicle ${this.vehicleId}`);
      this.onConnect?.();
    };

    this.ws.onmessage = (event) => {
      try {
        const data = JSON.parse(event.data);
        
        if (data.type === 'anomaly_alert' && data.data) {
          this.onAnomaly?.(data.data as AnomalyDetectionResult);
        } else if (data.type === 'connected') {
          console.log('Anomaly detection stream connected');
        } else if (data.type === 'keepalive') {
          // Keepalive - no action needed
        } else if (data.type === 'pong') {
          // Pong response - no action needed
        }
      } catch (error) {
        console.error('Error parsing WebSocket message:', error);
      }
    };

    this.ws.onerror = (error) => {
      console.error('Anomaly WebSocket error:', error);
      this.onError?.(error);
    };

    this.ws.onclose = () => {
      console.log(`Anomaly WebSocket disconnected for vehicle ${this.vehicleId}`);
      this.onDisconnect?.();
    };
  }

  disconnect() {
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
  }

  sendPing() {
    if (this.ws && this.ws.readyState === WebSocket.OPEN) {
      this.ws.send('ping');
    }
  }

  isConnected(): boolean {
    return this.ws !== null && this.ws.readyState === WebSocket.OPEN;
  }
}

File: docs/backend-development/pitwall-backend/app/models/anomaly.py
"""
Pydantic models for anomaly detection
"""
from pydantic import BaseModel, Field
from typing import List, Optional, Literal
from datetime import datetime


class AnomalyAlert(BaseModel):
    """Individual anomaly alert"""
    type: str = Field(..., description="Alert type: 'critical', 'rate_of_change', or 'ml_detected'")
    sensor: str = Field(..., description="Sensor name that triggered the alert")
    value: Optional[float] = Field(None, description="Sensor value (if applicable)")
    threshold: Optional[float] = Field(None, description="Threshold that was exceeded")
    rate: Optional[float] = Field(None, description="Rate of change (if applicable)")
    score: Optional[float] = Field(None, description="Anomaly score (0-1, if applicable)")
    contributing_features: Optional[List[str]] = Field(None, description="Features contributing to anomaly")
    message: str = Field(..., description="Human-readable alert message")
    severity: Literal["low", "medium", "high"] = Field(..., description="Alert severity level")


class AnomalyDetectionResult(BaseModel):
    """Result of anomaly detection on a single telemetry point"""
    is_anomaly: bool = Field(..., description="Whether any anomaly was detected")
    anomaly_score: float = Field(..., description="ML anomaly score (0-1, higher = more anomalous)")
    alerts: List[AnomalyAlert] = Field(default_factory=list, description="List of anomaly alerts")
    timestamp: str = Field(..., description="Timestamp of the telemetry point")
    vehicle_id: str = Field(..., description="Vehicle identifier")
    vehicle_number: Optional[int] = Field(None, description="Vehicle number")
    lap: Optional[int] = Field(None, description="Current lap number")


class AnomalyDetectionRequest(BaseModel):
    """Request for anomaly detection"""
    vehicle_id: str = Field(..., description="Vehicle identifier")
    telemetry_point: dict = Field(..., description="Telemetry data point as dictionary")


class AnomalyStats(BaseModel):
    """Statistics about anomalies detected"""
    total_points: int = Field(..., description="Total telemetry points analyzed")
    anomaly_count: int = Field(..., description="Number of points with anomalies")
    anomaly_rate: float = Field(..., description="Percentage of points with anomalies")
    critical_alerts: int = Field(..., description="Number of critical threshold alerts")
    rate_of_change_alerts: int = Field(..., description="Number of rate-of-change alerts")
    ml_detected_anomalies: int = Field(..., description="Number of ML-detected anomalies")
    avg_anomaly_score: float = Field(..., description="Average anomaly score")
    top_anomalous_sensors: List[dict] = Field(default_factory=list, description="Sensors with most anomalies")


class BatchAnomalyDetectionRequest(BaseModel):
    """Request for batch anomaly detection"""
    vehicle_id: str = Field(..., description="Vehicle identifier")
    track: str = Field(..., description="Track identifier")
    race: int = Field(..., description="Race number")
    lap_start: Optional[int] = Field(None, description="Starting lap")
    lap_end: Optional[int] = Field(None, description="Ending lap")
    retrain: bool = Field(True, description="Whether to retrain model on the data")


